cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java

```diff
@@ -0,0 +1,555 @@
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.Type;
+import org.objectweb.asm.commons.AdviceAdapter;
+import org.objectweb.asm.commons.GeneratorAdapter;
+import org.objectweb.asm.commons.Method;
+import org.objectweb.asm.tree.AnnotationNode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.Modifier;
+import java.util.ArrayList;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * <p>
+ * A {@link ClassRewriter} for rewriting bytecode of classes which needs Authorization Enforcement and uses
+ * {@link AuthEnforce} annotation.
+ * </p>
+ * <p>
+ * This class rewriter does two passes on the class:
+ * </p>
+ * <p>
+ * Pass 1: In first pass it visit all the classes skipping interfaces and looks for non-static methods which has
+ * {@link AuthEnforce} annotation. If an {@link AuthEnforce} annotation is found then it store all the
+ * {@link AuthEnforce} annotation details and also visits the parameter annotation and collects all the position of
+ * parameters with {@link Name} annotation. Note: No byte code modification of the class is done in this pass.
+ * </p>
+ * <p>
+ * Pass 2: This pass only happen if a method with {@link AuthEnforce} annotation is found in the class during the
+ * first pass. In this pass we rewrite the method which has {@link AuthEnforce} annotation. In the rewrite we
+ * generate a call to
+ * {@link AuthEnforceUtil#enforce(AuthorizationEnforcer, EntityId, AuthenticationContext, Set)} with
+ * all the annotation details collected in the first pass.
+ * </p>
+ * <p>
+ * Below is a sample of how a class containing a method annotation with {@link AuthEnforce} looks like before and
+ * after class rewrite.
+ * </p>
+ * <p>
+ * Before:
+ * <pre>
+ *     public class ValidAuthEnforceAnnotations {
+ *      private final AuthorizationEnforcer authorizationEnforcer;
+ *      private final AuthenticationContext authenticationContext;
+ *
+ *      public ValidAuthEnforceAnnotations() {
+ *        authenticationContext = new AuthenticationTestContext();
+ *        authorizationEnforcer = new ExceptionAuthorizationEnforcer();
+ *      }
+ *
+ *      &#064;AuthEnforce(entities = "namespaceId", enforceOn = NamespaceId.class, actions = Action.ADMIN)
+ *      public void testSingleAction(@Name("namespaceId") NamespaceId namespaceId) throws Exception {
+ *        System.out.println("Hello");
+ *      }
+ *     }
+ *   </pre>
+ * </p>
+ * <p>
+ * After:
+ * <pre>
+ *     public class ValidAuthEnforceAnnotations {
+ *      private final AuthorizationEnforcer authorizationEnforcer;
+ *      private final AuthenticationContext authenticationContext;
+ *
+ *      public ValidAuthEnforceAnnotations() {
+ *        authenticationContext = new AuthenticationTestContext();
```

Comment:
The setting should be done through `@Inject` setter method, not in the constructor.